docs/src/intro/usage.md

```diff
@@ -51,4 +52,138 @@ In the `miden/examples/fib` directory, we provide a very simple Fibonacci calcul
+
+### Miden assembly instruction
+
+All the Miden instruction mentioned in the Miden Assembly document is valid. (Ex. push.0, drop, dropw, swap, cswap, u32checked_add, mem_loadw, mtree_get etc.). One can either input instructions one by one or multiple instructions in one input.
+
+For example the below two commands will result in the same output.
+```
+>> push.1
+>> push.2
+>> push.3
+```
+
+```
+push.1 push.2 push.3
+```
+
+In order to execute a control flow operation, one needs to write the entire flow operation in a single line with spaces between individual operations.
+
+```
+repeat.20
+    pow2
+end
+```
+
+The above example should be written as following in the Repl tool:
+
+```
+repeat.20 pow2 end
+```
+
+### !help
+
+`!help` command prints out the list of all the available commands in the repl tool. 
+
+### !program
+
+`!program` command prints out the entire miden program being executed so far. For e.g. in the below scenario:
+
+```
+>> push.1.2.3.4
+>> repeat.16 pow2 end
+>> u32checked_add
+
+>> !program
+begin
+    push.1.2.3.4
+    repeat.16 pow2 end
+    u32checked_add
+end
+```
+
+### !stack
+
+`!stack` command prints out the state of the stack at the latest clock cycle. The state also contains the overflow part of the stack. The stack will be empty at initialisation i.e. it will only have 16 `0`. 
+
+```
+>> push.1 push.2 push.3 push.4 push.5
+>> exp
+>> u32checked_mul
+>> swap
+>> eq.2
+>> assert
+```
+
+The `!stack` command will print out the following state of the stack:
+```
+>> !stack
+3072 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
+```
+
+### !mem
+
+`!mem` command prints the entire state of the memory which has been initialised. For all the values memory addresses which has been initialised, it prints the address along with it's memory register's value (integer).
+
+If the memory has at least one value which has been initialised:
+
+```
+>> !mem
+7: [1, 2, 0, 3]
+8: [5, 7, 3, 32]
+9: [9, 10, 2, 0]
+```
+
+Here 7, 8 and 9 addresses have been initialised and their respective counter-parts in the same line are the memory values on these addresses.
+
+If the memory is not yet been initialised, then the commands prints that the memory is not yet initialsed.
+
+```
+> !mem
+The memory has not been initialised yet
+```
+
+### !mem[addr]
+
+`!mem[addr]` prints the memory value(in integer) present in the address `addr`. 
+
+If the `addr` has been initialied:
+
+```
+>> `!mem[9]`
+9: [9, 10, 2, 0]
+```
+
+If the `addr` has not been initialied:
+
+```
+>> !mem[87]
+Memory at address 87 is empty
+```
+
+### !undo
+
+`!undo` command reverts to the previous state of the stack and memory by dropping off the last executed assembly instruction from the program. One could use `undo` as many times as they want to restore the state of a stack and memory n instructions ago(provided there are n instructions in the program). If there are no instructions in the repl tool and `!undo` command is sent as an input then it will result in an error. 
```

Comment:
* Missing space after "ago".
* Let's write `n` as `$n$` so that it appears in math notation.
* In one case `undo` does not start with `!`.

Reply:
Fixed, thanks!